File: auth_service/src/domain/entities/message.ts
import { ID } from "./OV/id";

export interface MessageDTO {
    id?: string;
    senderId: string;
    content: string;
    translatedContent: string;
    timestamp: string;
    status: MessageStatus;
}

export enum MessageStatus {
    SENT = "SENT",
    RECEIVED = "RECEIVED",
    READ = "READ"
}

export class Message {
    public id: ID;
    public senderId: ID;
    public content: string;
    public translatedContent: string;
    public timestamp: string;
    public status: MessageStatus;

    constructor(dataMessage: MessageDTO){
        this.id = new ID(dataMessage.id);
        this.senderId = new ID(dataMessage.senderId);
        this.content = dataMessage.content;
        this.timestamp = dataMessage.timestamp;
        this.status = dataMessage.status;
        this.translatedContent = dataMessage.translatedContent;
    }

    public toMongo(): MessageDTO {
        return {
            senderId: this.senderId.getValue,
            content: this.content,
            timestamp: this.timestamp,
            status: this.status,
            translatedContent: this.translatedContent
        };
    }

    public toNotification(): MessageDTO{
        return {
            id: this.id.getValue,
            content: this.content,
            translatedContent: this.translatedContent,
            senderId: this.senderId.getValue,
            status: this.status,
            timestamp: this.timestamp
        }
    }
}

File: chat_service/lib/chat_service/models/message.ex
defmodule ChatService.Models.Message do
  @enforce_keys [:id, :sender_id, :content, :timestamp, :status]
  defstruct id: "",
            sender_id: "",
            content: "",
            translated_content: "",
            timestamp: "",
            status: ""
            # mobile_ref_id: ""

  @type t :: %__MODULE__{
          id: String.t(),
          sender_id: String.t(),
          content: String.t(),
          translated_content: String.t(),
          timestamp: String.t(),
          status: String.t(),
          # mobile_ref_id: String.t()
        }

  def to_map(%__MODULE__{} = message) do
    Map.from_struct(message)
  end
end
